document serving's revision garbage collection (#4266)
knative/docs: docs/admin/serving/revision-gc.md

> # Revision Garbage Collection

Reviewer: ```suggestion
# Revision garbage collection
```

> The `config-gc` ConfigMap contains settings that determine how to clean up in-active revisions. This ConfigMap is located in the `knative-serving` namespace.

Reviewer: ```suggestion
The `config-gc` ConfigMap contains settings that determine how inactive revisions are cleaned up. This ConfigMap is located in the `knative-serving` namespace.
```

> ## Cluster-wide Configuration

Reviewer: ```suggestion
## Cluster-wide configuration
```

name|description
`retain-since-create-time`| Duration since creation before considering a revision for GC or "disabled"
`retain-since-last-active-time`| Duration since active before considering a revision for GC or "disabled"
`min-non-active-revisions`| Minimum number of non-active revisions to retain.
`max-non-active-revisions`| Maximum number of non-active revisions to retain or "disabled" to disable any maximum limit.

Revisions are retained if they fall into any one of the following categories
- Is active and is being reference by a route
- Created within `retain-since-create-time`
- Last referenced by a route within `retain-since-last-active-time`
- There are fewer than `min-non-active-revisions`

### Examples

#### Immediately collect any inactive revision

```
min-non-active-revisions: "0"
retain-since-create-time: "disabled"
retain-since-last-active-time: "disabled"
```

#### Keep around the last ten non-active revisions
```
retain-since-create-time: "disabled"
retain-since-last-active-time: "disabled"
max-non-active-revisions: "10"
```

#### Disable GC
```
retain-since-create-time: "disabled"
retain-since-last-active-time: "disabled"
max-non-active-revisions: "disabled"

> ```

Reviewer: can you show some context for these, i.e. where in the configmap would they appear? I feel like these might be a bit too abstract as examples

> #### Complex Example

Reviewer: ```suggestion
#### Complex example
```

> This example config wll keep recently deployed or active revisions,
> always maintain the last two in case of rollback, and prevent burst
> activity from exploding the count of old revisions.

Reviewer: ```suggestion
The following example configuration keeps recently deployed or active revisions,
always maintains the last two revisions in case of rollback, and prevents burst
activity from exploding the count of old revisions:
```

```
retain-since-create-time: "48h"
retain-since-last-active-time: "15h"
min-non-active-revisions: "2"
max-non-active-revisions: "1000"

> ```

Reviewer: Please show more context as per above comment - this doesn't show where in the configmap these would be added etc

> ## Per Revision Options

Reviewer: ```suggestion
## Per-revision options
```

> Users can alwaysretain a revision by adding the annotation `serving.knative.dev/no-gc: "true"`

Reviewer: ```suggestion
Users can configure their settings so that at least one revision is always retained, by adding the `serving.knative.dev/no-gc: "true"` annotation to a Service:
```

# Revision garbage collection

The `config-gc` ConfigMap contains settings that determine in-active revisions are cleaned up. This ConfigMap is located in the `knative-serving` namespace.

## Cluster-wide configuration

The following properties allow you to configure revision garbage collection:

-|-
name|description
`retain-since-create-time`| Duration since creation before considering a revision for GC or "disabled"
`retain-since-last-active-time`| Duration since active before considering a revision for GC or "disabled"
`min-non-active-revisions`| Minimum number of non-active revisions to retain.

> `max-non-active-revisions`| Maximum number of non-active revisions to retain or "disabled" to disable any maximum limit.

Reviewer: Presumably this should be a table but I think the formatting is wrong since it's not rendering in the preview?
https://deploy-preview-4266--knative.netlify.app/development/admin/serving/revision-gc/#cluster-wide-configuration

Author: good catch - let me fix that

> Revisions are retained if they fall into any one of the following categories
> - Is active and is being reference by a route
> - Created within `retain-since-create-time`
> - Last referenced by a route within `retain-since-last-active-time`
> - There are fewer than `min-non-active-revisions`
> 

Reviewer: ```suggestion
Revisions are retained if they fall into any one of the following categories:

- Is active and is being reference by a route
- Created within `retain-since-create-time`
- Last referenced by a route within `retain-since-last-active-time`
- There are fewer than `min-non-active-revisions`
```

Fix line spacing so this renders properly in the preview

### Examples

#### Immediately collect any inactive revision
```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: config-gc
  namespace: knative-serving
data:
  min-non-active-revisions: "0"
  retain-since-create-time: "disabled"
  retain-since-last-active-time: "disabled"
```

#### Keep around the last ten non-active revisions
```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: config-gc
  namespace: knative-serving
data:
  retain-since-create-time: "disabled"
  retain-since-last-active-time: "disabled"
  max-non-active-revisions: "10"
```

> #### Disable GC

Reviewer: ```suggestion
#### Disable garbage collection
```
I'd spell this out tbh

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: config-gc
  namespace: knative-serving
data:
  retain-since-create-time: "disabled"
  retain-since-last-active-time: "disabled"
  max-non-active-revisions: "disabled"
```

#### Complex example

> Th following example configuration keeps recently deployed or active revisions,
> always maintains the last two revisions in case of rollback, and prevents burst
> activity from exploding the count of old revisions:

Reviewer: ```suggestion
The following example configuration keeps recently deployed or active revisions,
always maintains the last two revisions in case of rollback, and prevents burst
activity from exploding the count of old revisions:
```
Minor typo

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: config-gc
  namespace: knative-serving
data:
  retain-since-create-time: "48h"
  retain-since-last-active-time: "15h"
  min-non-active-revisions: "2"
  max-non-active-revisions: "1000"
```

## Per-revision options

> You can always retain a revision by adding the `serving.knative.dev/no-gc: "true"` annotation

Reviewer: ```suggestion
You can configure a revision so that it is never garbage collected by adding the `serving.knative.dev/no-gc: "true"` annotation:
```
I'd reword this to make it clearer - the always retain bit just reads weird to me. Might be just me though.
